File: 0496-next-greater-element-i/0496-next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        # Brute Force Solution: To find the matching values first and then
        # find the next greater element for that element;

        nums1_len = len(nums1)
        next_greater_els = [-1] * nums1_len

        idx_i = 0
        for num in nums1:
            for idx_j in range(len(nums2)):
                second_num = nums2[idx_j]

                if num == second_num:
                    idx_j += 1
                    while (idx_j < len(nums2) and second_num > nums2[idx_j]):
                        idx_j += 1
                    
                    if idx_j < len(nums2):
                        next_greater_els[idx_i] = nums2[idx_j]
                    else:
                        next_greater_els[idx_i] = -1
                    
                    idx_i += 1
                    break
            
        return next_greater_els

        '''
        # if nums1 is a subset of nums2 then definitely
        # all the elements of nums1 are present in nums2:

        # Logic: Find the next greater element for each element of 
        # nums1 and keep it thier answer in their indices in the results
        # to check if the answer needs to be updated for nums1:
        # we can check if it is present in nums1 and at what index?
        nums1_len = len(nums1)

        nums1_els_map = {}
        next_greater_stack = []
        next_greater_els = [-1] * nums1_len

        for idx_i in range(nums1_len):
            num = nums1[idx_i]
            nums1_els_map[num] = idx_i
        
        for idx_j in range(len(nums2) - 1, -1, -1):
            while next_greater_stack and next_greater_stack[-1] <= nums2[idx_j]:
                next_greater_stack.pop()
            
            if next_greater_stack and nums2[idx_j] in nums1_els_map:
                idx_value = nums1_els_map.get(nums2[idx_j])

                next_greater_els[idx_value] = next_greater_stack[-1]
            
            next_greater_stack.append(nums2[idx_j])
        
        return next_greater_els
        '''
        
        

File: 0496-next-greater-element-i/0496-next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        HashMap <Integer, Integer> nums2GreaterElsMap = new HashMap<>();

        int n = nums2.length;

        int[] nextGreaterEls = new int[nums1.length];
		
		Deque<Integer> nextGreaterStack = new ArrayDeque<>();
		
		for(int idx = n - 1; idx >= 0; idx--){
			int currEl = nums2[idx];
			
			while(!nextGreaterStack.isEmpty() && nextGreaterStack.peek() <= currEl){
				nextGreaterStack.pop();
			}
			
			if(!nextGreaterStack.isEmpty()){
				nums2GreaterElsMap.put(currEl, nextGreaterStack.peek());
			}
			else{
				nums2GreaterElsMap.put(currEl, -1);
			}
			
			nextGreaterStack.push(currEl);
		}

        // int idx = 0;
        // for(int num : nums1){
        //     // nums1 is a subset so no need of else & if;
        //     if(nums2GreaterElsMap.containsKey(num)){
        //         nextGreaterEls[idx++] = nums2GreaterElsMap.get(num);
        //     }
        // }

        // OR;
        int idx = 0;
        for(int num : nums1){
            // nums1 is a subset so no need of else & if;
            nextGreaterEls[idx++] = nums2GreaterElsMap.get(num);
        }
		
		return nextGreaterEls;

        // Dry Run:
        // nums2Map = {2: -1, 4: -1, 3: 4, 1: 3}
        // nums2Stack = []
        // nums2Stack = [2] => [] then 4 
        // numsStack = [4] => [4, 3] => [4, 3, 1]
        // nums1 => 4 in map => ans is -1,
        // nums1 => 1 in map => ans is 3
        // nums1 => 2 in map => ans is -1
    }
}